### ISSUE TITLE
[BUG] Double runs created using databricks projects backend, `_mlflow_local_backend_run_id` seems ignored.

### ISSUE THREAD

**Issue by FlorisHoogenboom** [mlflow/mlflow#3787]
### Willingness to contribute
The MLflow Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the MLflow code base?

- [x] Yes. I can contribute a fix for this bug independently.
- [ ] Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.
- [ ] No. I cannot contribute a bug fix at this time.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 20.04
- **MLflow installed from (source or binary)**: source
- **MLflow version (run ``mlflow --version``)**: 1.12
- **Python version**: 3.7
- **npm version, if running the dev UI**: not applicable
- **Exact command to reproduce**: `mlflow run . -b databricks --backend-config cluster-config.json`

### Describe the problem
When running mlflow projects using the Databricks projects backend two distinct run_id's seem to be created. The first run id is created by the client submitting the shell command job to databricks. This creates a run on the tracking server and connects the job status to that run. The MLFlow library then creates a bash command job as a databricks job by running the following command on the databricks job

```
mlflow run PATH_ON_DATABRICKS --entry-point MY_ENTRY_POINT
 -c '{"_mlflow_local_backend_run_id": "THE RUN ID OF THE RUN CREATED BY THE CLIENT"}'
```

However, when this command is executed in the databricks job the config for the local backend passed by `_mlflow_local_backend_run_id` seems to be ignored. This results in a new run being created on the tracking server. This is the run where the databricks job will log all its parameters to.

It seems easy to fix this issue by, instead of passing the client run id to the databricks job via that (seemingly private) `_mlflow_local_backend_run_id` parameter as the backend config, using the `--run-id` cli option. I've created a small MR to show how this could be resolved (#3788 ).

### Code to reproduce issue
I am not able to reproduce this issue locally, it seems only in the databricks job the local backend config `_mlflow_local_backend_run_id` is ignored.

To reproduce, create any simple mlflow project and try to run that with the databricks backend.

### Other info / logs
[This](https://ibb.co/9w2Fjny) is a picture of how runs show up in the UI. The bottom run is created by the client running `mlflow run . -b databricks` the top run is the run created by the bash job executed on databricks that actually performs the work. The bottom run will always have state UNFINISHED since it is never completed by any running job.

### What component(s), interfaces, languages, and integrations does this bug affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [x] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [x] `integrations/databricks`: Databricks integrations

**Comment by FlorisHoogenboom** [comment]
Me and my colleague @SebGr did a bit more investigation on this issue and it seems that this was introduced in `1.9`. This is also the version where `mlflow` started using the local backend config to pass the `run_id` to the databricks job instead of relying on the `--run-id` command line argument (as I'm proposing to as a fix in #3788).

**Comment by AveshCSingh** [comment]
@FlorisHoogenboom thank you for filing this issue. I am having trouble reproducing when using a Databricks backend. Would you mind providing the following:

a) Version of mlflow specified in your project's `conda.yaml`.
b) [Databricks runtime version](https://docs.databricks.com/release-notes/runtime/releases.html) (eg "Databricks Runtime 7.5 for ML")

**Comment by FlorisHoogenboom** [comment]
@AveshCSingh . Thank you for looking into this. We we're running still on 6.6-ml (needing to upgrade to 7.3-ml-lts). As it turns out the issue only exists when running on 6.6-ml and not on 7.3-ml. I've created a small repo for you to reproduce. 

See [here](https://github.com/FlorisHoogenboom/mlflow-double-runs/tree/master) to reproduce the issue on 6.6-ml. See here to run the same project on [7.3-ml-lts ](https://github.com/FlorisHoogenboom/mlflow-double-runs/tree/dbr7.3)(which does not create double runs). Both use `mlflow=1.12.1` in their `environment.yaml` file ([here](https://github.com/FlorisHoogenboom/mlflow-double-runs/blob/4ed0cae5a04579441d1a037d7e44343cafe16260/mlflow-environment.yml#L8)).

It is interesting to see it does not appear on `7.3-ml`. To me that seems to point how that local backend config is interpreted on the backend.

**Comment by AveshCSingh** [comment]
Ah I see. The problem is that these earlier versions of the ML runtime run mlflow <= 1.8.0. The backend_config which specifies `_MLFLOW_LOCAL_BACKEND_RUN_ID_CONFIG` is not backwards compatible to these earlier versions of MLflow. One possible fix is to both pass the backend config as well as `--run-id`.

If you would like to implement this, I would be happy to review. Otherwise, I can see if the engineering team at Databricks has the bandwidth.

**Comment by FlorisHoogenboom** [comment]
Ahh, that explains a lot. Thank you so much for helping out here. We've been banging our head against this one for quite some time! I do feel that this is actually quite a tricky thing: the MLFlow version to start the run using the local backend is not the version specified in your projects `conda.yaml` or the version specified as the cluster dependency (as part of the submitted job). This now means that it is very easy to end up with a project where you have three different versions of MLFlow:

- The version of your client
- The version you specify in your projects `conda.yaml` and hence use to log
- The version that comes preinstalled on the databricks runtime.

I do think that this might be something that would be good to document a bit more (happy to submit an MR for that if you agree here), but also it might be better to just pin the version of MLFlow that is installed on the cluster level. Now if there is by default an older version of MLFlow on the runtime that is allowed, but I think it might be better to just pin the version to the same version as the client here: https://github.com/mlflow/mlflow/blob/c7b5c5dd3ab7281f6fd9b0d44f78589353989b46/mlflow/projects/databricks.py#L199

That solves the issue of the "outer" MLFlow on databricks running a different version than the local client you are running locally and hence would prevent similar issues like this one in the future!

**Comment by AveshCSingh** [comment]
Thanks for summarizing this so clearly! I think using the same version for `(2) The version you specify in your projects conda.yaml and hence use to log` and `(3) The version that comes preinstalled on the databricks runtime.` would reduce the cognitive overhead here.

Unfortunately, we would need this change to be backwards compatible. Meaning that MLProjects running against existing Databricks runtimes should continue to operate as-is, while new runtimes can fix the version of mlflow used in (3).

Another worry is that we may be breaking some use cases. @dbczumar @juntai-zheng @smurching is it common for a user to run an MLflow project against an MLflow backend, but want to use a different version of MLflow than that backend comes with?

**Comment by AveshCSingh** [comment]
I spoke with @smurching. Pinning the version of mlflow here
https://github.com/mlflow/mlflow/blob/c71bae2365177e57d7b50004c5b9964bd455bacc/mlflow/projects/databricks.py#L199
as you suggested sounds good. Though we will need to add some special logic to work around the issue described in that comment. Something like this pseudo-code:
```
if VERSION.ends_with('.dev0'):
  libraries = [{"pypi": {"package": "'mlflow<=%s'" % VERSION}}] 
else:
  libraries = [{"pypi": {"package": "'mlflow=%s'" % VERSION}}] 
```

**Comment by FlorisHoogenboom** [comment]
Perfect! I've created an MR containing an implementation of these changes: #3880.

**Comment by AveshCSingh** [comment]
Wonderful! Thank you for fixing this point of confusion. I've LGTM'd and merged your pull request.

The PR auto-closed this Issue, but I'm actually going to re-open it. I believe that we still need to fix the run_id issue, perhaps by passing the backend config as well as --run-id.

**Comment by FlorisHoogenboom** [comment]
Thnx! I do think that this specific issue is resolved by #3880 (at least when running a non-edge version of the client), since that forces the same version of MLFlow to run on the remote as well as the submitting client. Issues may still occur with an (older) databricks runtime and an edge version of the client (e.g. by installing from master), but in those cases a user will now be warned.

I do however still think that is would be nicer to use use the `--run-id` parameter of the CLI to comunicate that run_id to the databricks job, rather than relying on the `_MLFLOW_LOCAL_BACKEND_RUN_ID_CONFIG`. I am still missing the purpose of using that mechanism to communicate. It seems this was introduced in #2566, but from the discussion there I also cannot see a clear reason why it was necessary to replace the "old" way of passing these arguments (which is via `--run-id`) needed to be superseeded by that `_MLFLOW_LOCAL_BACKEND_RUN_ID_CONFIG`.

Anyways, going forward we basically have three options:
- Only use `--run-id` for passing these parameters to the remote. That's what #3788 proposed. In that case I do think we should get a bit of insight into why we switched away from this previously. Maybe @jdlesage can provide some perspective here?
- Only use the `_MLFLOW_LOCAL_BACKEND_RUN_ID_CONFIG` for passing that run id. That's keeping the status quo and will only produce issues when using an older databricks backend and an edge version of the CLI client.
- Use both. I'm personally not a big fan of that since it may be harder to maintain. It requires anybody making changes to that part in the future to be aware of that there are two ways those run ids are passed and I cannot oversee what would happen if you pass e.g. a different run id via the `--run-id` cli parameter and one via the `_MLFLOW_LOCAL_BACKEND_RUN_ID_CONFIG`.

Personally, I would prefer option 1 since it uses the public API and seems the most clear.

**Comment by AveshCSingh** [comment]
You're right; the run_id issue is resolved by #3880. 

As for `--run-id` vs `_MLFLOW_LOCAL_BACKEND_RUN_ID_CONFIG`, I'm also curious to hear @jdlesage and @smurching 's thoughts.

**Comment by FlorisHoogenboom** [comment]
I think this has been stale for a while and the issue is resolved by the ML. I'm gonna close this but if anyone of you wants to restart the discussion on the `_MLFLOW_LOCAL_BACKEND_RUN_ID_CONFIG` vs `--run-id` feel free to reopen.
